File: src/main/java/com/indiepost/service/CategoryServiceImpl.java
package com.indiepost.service;

import com.indiepost.dto.CategoryDto;
import com.indiepost.model.Category;
import com.indiepost.repository.CategoryRepository;
import org.springframework.beans.BeanUtils;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import javax.inject.Inject;
import java.util.List;
import java.util.stream.Collectors;

/**
 * Created by jake on 8/4/16.
 */
@Service
@Transactional
public class CategoryServiceImpl implements CategoryService {

    private final CategoryRepository categoryRepository;

    @Inject
    public CategoryServiceImpl(CategoryRepository categoryRepository) {
        this.categoryRepository = categoryRepository;
    }

    @Override
    public void save(Category category) {
        categoryRepository.save(category);
    }

    @Override
    public void update(Category category) {
        categoryRepository.update(category);
    }

    @Override
    public void delete(Category category) {
        categoryRepository.delete(category);
    }

    @Override
    public Category getReference(Long id) {
        return categoryRepository.getReference(id);
    }

    @Override
    public Category findById(Long id) {
        return categoryRepository.findById(id);
    }

    @Override
    public Category findBySlug(String slug) {
        return categoryRepository.findBySlug(slug);
    }

    @Override
    public List<Category> findAll() {
        return categoryRepository.findAll();
    }

    @Override
    public List<Category> findByParent(Category parent) {
        return categoryRepository.findByParentId(parent.getId());
    }

    @Override
    public List<CategoryDto> getDtoList() {
        return this.findAll()
                .stream()
                .map(category -> {
                    CategoryDto categoryDto = new CategoryDto();
                    BeanUtils.copyProperties(category, categoryDto);
                    return categoryDto;
                })
                .collect(Collectors.toList());
    }
}

File: src/main/kotlin/com/indiepost/service/CategoryServiceImpl.kt
package com.indiepost.service

import com.indiepost.dto.CategoryDto
import com.indiepost.model.Category
import com.indiepost.repository.CategoryRepository
import org.springframework.beans.BeanUtils
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.util.stream.Collectors
import javax.inject.Inject

/**
 * Created by jake on 8/4/16.
 */
@Service
@Transactional
class CategoryServiceImpl @Inject
constructor(private val categoryRepository: CategoryRepository) : CategoryService {

    override fun save(category: Category) {
        categoryRepository.persist(category)
    }

    override fun delete(category: Category) {
        categoryRepository.delete(category)
    }

    override fun getReference(id: Long): Category? {
        return categoryRepository.getReference(id)
    }

    override fun findById(id: Long): Category? {
        return categoryRepository.findById(id)
    }

    override fun findBySlug(slug: String): Category? {
        if (slug.isEmpty()) {
            return null
        }
        return categoryRepository.findBySlug(slug)
    }

    override fun findAll(): List<Category> {
        return categoryRepository.findAll()
    }

    override fun findByParent(parent: Category): List<Category> {
        return categoryRepository.findByParentId(parent.id!!)
    }

    override fun getDtoList(): List<CategoryDto> {
        return this.findAll()
                .stream()
                .map { category ->
                    val categoryDto = CategoryDto()
                    BeanUtils.copyProperties(category, categoryDto)
                    categoryDto
                }
                .collect(Collectors.toList())
    }
}
